### PR TITLE
MSC3884: Sliding Sync Extension: E2EE

### PR DESCRIPTION
**Pull request by kegsay**
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/kegan/msc3884/proposals/3884-sliding-sync-e2ee.md)

[Client Implementation](https://github.com/matrix-org/matrix-js-sdk/blob/v19.5.0-rc.4/src/sliding-sync-sdk.ts#L44)
[Server Implementation](https://github.com/matrix-org/sliding-sync/blob/main/sync3/extensions/e2ee.go)

### CODE REVIEW

File: proposals/3884-sliding-sync-e2ee.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+{
+    "device_one_time_keys_count": {
+        "signed_curve25519": 50
+    },
+    "device_lists": {
+        "changed": ["@alice:example.com"],
+        "left": ["@bob:example.com"]
+    },
+    "device_unused_fallback_key_types": [
+        "signed_curve25519"
```
Comment (deepbluev7) [inline proposals/3884-sliding-sync-e2ee.md:29]: This field has to always be included, otherwise there is no way for the client to tell, if the fallback key got used.

HOWEVER:

You can actually probably solve that with the sliding sync API, by having the client explicitly tell the server about what fallback keys it wants to know about, when they are **used**. Then the server would have to include them everytime when they are used, but could not include them in the common case of them being unused. The same could be done for the one time keys count, where you specify a minimum number you want to have and if you have those, the server doesn't update you about it.

This is however a bit racy, as a client might be told about a used key, while it is still uploading a key. But I think that is already an issue we have today.

Basically the request parameters would then look something like this:

```json5
{
  "enabled": true,
  "minimum_device_one_time_keys_count": {
    "signed_curve25519": 50 // tell the server to notify the client, if the OTKs fall below this limit
  },
  "fallback_key_types": [ "signed_curve25519" ] // tell the server to notify the client, if no such key is present
}
```

File: proposals/3884-sliding-sync-e2ee.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+```
+If `enabled` is `true`, then the sliding sync response MAY include the following response fields in
+the `e2ee` extension response:
+```json
+{
+    "device_one_time_keys_count": {
+        "signed_curve25519": 50
+    },
+    "device_lists": {
+        "changed": ["@alice:example.com"],
```
Comment (deepbluev7) [inline proposals/3884-sliding-sync-e2ee.md:25]: What are the semantics for this? When do I get a "changed" event? Only for device lists that changed during this connection? Or does this work on the delta token? Maybe this should have its own since and next batch token? Key data is actually a LOT of data, so invalidating that on every reconnect would be pretty scary. And not being told your device list is possibly missing changed events, because the delta token expired is equally scary.

File: proposals/3884-sliding-sync-e2ee.md (hunk 1 above)
Comment (deepbluev7) [reply proposals/3884-sliding-sync-e2ee.md:29]: This probably requires the server to tell the client, if a key type is unsupported though.

Comment (deepbluev7) [reply proposals/3884-sliding-sync-e2ee.md:29]: Related spec issue: https://github.com/matrix-org/matrix-spec-proposals/issues/3298

Unsupported keys are hard to distinguish from "no change".

File: proposals/3884-sliding-sync-e2ee.md (hunk 2 above)
Comment (kegsay) [reply proposals/3884-sliding-sync-e2ee.md:25]: The semantics are the same as sync v2 regarding getting a "changed" or "left" event.

Currently, the proxy will aggregate changed/left responses from sync v2 then deliver this to the client, only purging the arrays server-side once the client has acknowledged the response. This does mean that if you're offline for a while, these arrays can be large. We don't currently have delta token support in this extension, but it might be worthwhile to add it if we can get the semantics right.

Also worth noting that in practice the size of these arrays isn't a large part of the overall sync response, even if you have been offline for a while.

File: proposals/3884-sliding-sync-e2ee.md (hunk 1 above)
Comment (kegsay) [reply proposals/3884-sliding-sync-e2ee.md:29]: I like this idea overall, especially with the "min" values. I'd be interested to hear if this makes things any easier on clients, or just saves a bit of bandwidth.

File: proposals/3884-sliding-sync-e2ee.md (hunk 2 above)
Comment (deepbluev7) [reply proposals/3884-sliding-sync-e2ee.md:25]: My concern is mostly, that it would be easy to get this wrong on the client side, if you don't treat all devices as changed after a reconnection, where the server doesn't send you change lists from before you reconnected. Currently clients mostly rely on getting all change events as long as they don't do a new initial sync, but with sliding sync the "initial sync" could possibly happen quite often and by accident and in that case you would need to treat ALL keys as stale. You can probably implement that correctly in a client, but some more explicit notification, that your delta token wasn't used and your whole lists should be treated as stale might be harder to get wrong.

**Comment by erikjohnston** [comment]
Superseded by https://github.com/matrix-org/matrix-spec-proposals/pull/4543
